### DISCUSSION TITLE
Vocab is not getting stored for custom trained NER model

### DISCUSSION THREAD

**Discussion by davesuketu215** [explosion/spaCy#8108] [Help: Coding & Implementations] [answered]
I'm training a custom NER model using code approach from scratch, with tok2vec and ner pipes in the pipeline. Close to a million training records. v3.06
However, when I save the model using ```nlp.to_disk( )``` I do not see my tokens in vocab/strings.json.

As mentioned here - (https://github.com/explosion/spaCy/issues/3593#issuecomment-532203962), do I need to explicitly create a new vocab for my custom model? Is there any systematic documentation which can guide me step-by-step for the same?

However, I am surprised to see that when I load the model, it does give me predictions, even if the entire saved model folder is not more than 8 MB in size and missing tokens in strings.json

**Comment by davesuketu215** [comment]
In addition to this, just before performing ```nlp.to_disk( )``` when i print ```nlp.meta['vectors']``` I get the below output:
```{'width': 0, 'vectors': 0, 'keys': 0, 'name': None}```
But the saved model is working for making inferences.

**Reply by adrianeboyd** [reply]
Those are static vectors, not `tok2vec` vectors, which are really context-sensitive tensors.

If you look at `xx_ent_wiki_sm` as an example (trained on a huge amount of data), the NER component is ~7 MB and most of the size is the string store, which has strings from ~10 languages.

**Comment by adrianeboyd** [comment, marked as answer]
I'm not sure how you trained/saved your model so that it's not saving the same vocab that you trained with, but the NER model will work even if you remove all the strings from the string store.

When you process a new `doc`, its token strings are converted to integer hash values underneath, and those are what the NER model uses as input. It predicts from the integer hash values for `NORM`/`PREFIX`/etc. and doesn't need any stored string values to work correctly (aside from its own labels like `PERSON`, which are stored in the component data).

**Reply by davesuketu215** [reply]
btw, I had trained my model using the below snippet:
```
        batches = minibatch(examplesTrain, size=batchSize)
        for batch in batches:
            nlp.update(batch, sgd=optimizer, drop=dropoutRate losses=lossesTrain)
```

And to save:
```
nlp.to_disk(myPath')
```

Are the vocabs being list somewhere in the process? I am finding it really difficult to trace it.

**Reply by adrianeboyd** [reply]
You can check the size of the string store with `len(nlp.vocab.strings)` and run `nlp.vocab.to_disk()` just to save the vocab to try to figure out what's going on. I don't see anything unexpected in this snippet, but without the full code it's hard to to know what might be going on. Is it possible you've created the training docs/examples with a different pipeline?

**Reply by davesuketu215** [reply]
Yes, actually we have our own text data, using which I have created docs and examples. And those same examples I feed to the batches ```minibatch( )``` as ```examplesTrain```. And this operation is happening in the same service.

The flow goes like this -->
Raw Annotations --> Doc --> Span Adjustments --> Annotations with adjusted spans --> Doc --> Examples --> Feed to nlp.update( ) in batches
Is there any place where I need to explicitly feed the tokens into ```nlp.vocab```?

Also, on running ```len(nlp.vocab.strings)``` I get 1260, which is not what is expected.

**Reply by adrianeboyd** [reply]
The question is whether you do something like this:

```python
nlp = spacy.load("some_model")
# create Examples

nlp = spacy.blank("en") # not the same vocab!
nlp.add_pipe("blah")
# train
```

**Reply by davesuketu215** [reply]
Exactly the second one, ```spacy.blank("en")```
And I am very much doubtful of missing something in this process wrt vocab.

**Reply by adrianeboyd** [reply]
If everything trains correctly, it's not necessary to change this, but you can use the same vocab for the second pipeline like this, and then you should see all the strings in the string store:

```python
second_nlp = spacy.blank("en", vocab=first_nlp.vocab)
```

Again, you don't actually have to have all the strings for the model to work. `spacy train` does include all the strings by creating the examples and training with the same pipeline.
